File: cmd/lsif-server/src/shared/paths.ts
import * as constants from './constants'
import * as fs from 'mz/fs'
import * as path from 'path'

/**
 * Construct the path of the SQLite database file for the given dump.
 *
 * @param storageRoot The path where SQLite databases are stored.
 * @param id The ID of the dump.
 */
export function dbFilename(storageRoot: string, id: number): string {
    return path.join(storageRoot, constants.DBS_DIR, `${id}.lsif.db`)
}

/**
 * Construct the path of the raw upload file for the given identifier.
 *
 * @param storageRoot The path where uploads are stored.
 * @param id The identifier of the upload.
 */
export function uploadFilename(storageRoot: string, id: number): string {
    return path.join(storageRoot, constants.UPLOADS_DIR, `${id}.lsif.gz`)
}

/**
 * Returns the identifier of the database file. Handles both of the
 * following formats:
 *
 * - `{id}.lsif.db`
 * - `{id}-{repo}-{commit}.lsif.db`
 *
 * @param filename The filename.
 */
export function idFromFilename(filename: string): number | undefined {
    const id = parseInt(path.parse(filename).name.split('-')[0], 10)
    if (!isNaN(id)) {
        return id
    }

    return undefined
}

/**
 * Ensure the directory exists.
 *
 * @param directoryPath The directory path.
 */
export async function ensureDirectory(directoryPath: string): Promise<void> {
    try {
        await fs.mkdir(directoryPath)
    } catch (error) {
        if (!(error && error.code === 'EEXIST')) {
            throw error
        }
    }
}

File: cmd/precise-code-intel-bundle-manager/internal/paths/paths.go
package paths

import (
	"fmt"
	"os"
	"path/filepath"
)

// PrepDirectories
func PrepDirectories(bundleDir string) error {
	for _, dir := range []string{UploadsDir(bundleDir), DBsDir(bundleDir)} {
		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
			return err
		}
	}

	return nil
}

// UploadsDir returns the path of the uploads directory.
func UploadsDir(bundleDir string) string {
	return filepath.Join(bundleDir, "uploads")
}

// DBsDir returns the path of the dbs directory.
func DBsDir(bundleDir string) string {
	return filepath.Join(bundleDir, "dbs")
}

// UploadFilename returns the path fo the upload with the given identifier.
func UploadFilename(bundleDir string, id int64) string {
	return filepath.Join(bundleDir, "uploads", fmt.Sprintf("%d.lsif.gz", id))
}

func UploadPartFilename(bundleDir string, id, index int64) string {
	return filepath.Join(bundleDir, "uploads", fmt.Sprintf("%d.%d.lsif.gz", id, index))
}

// DBFilename returns the path fo the database with the given identifier.
func DBFilename(bundleDir string, id int64) string {
	return filepath.Join(bundleDir, "dbs", fmt.Sprintf("%d.lsif.db", id))
}

// PathExists returns (true, nil) if the specified path exists, or (false, error) if an error
// occurred (such as not having permission to read the path).
func PathExists(filename string) (bool, error) {
	if _, err := os.Stat(filename); err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}

		return false, err
	}

	return true, nil
}
